kubernetes/structures_deployment.go

```diff
@@ -90,40 +114,37 @@ func expandPodTemplate(l []interface{}) (*corev1.PodTemplateSpec, error) {
 }
 
 func expandDeploymentStrategy(l []interface{}) appsv1.DeploymentStrategy {
+	obj := appsv1.DeploymentStrategy{}
+
 	if len(l) == 0 || l[0] == nil {
-		return appsv1.DeploymentStrategy{}
+		obj.Type = appsv1.RollingUpdateDeploymentStrategyType
+		return obj
 	}
 	in := l[0].(map[string]interface{})
-	obj := appsv1.DeploymentStrategy{}
-	if v, ok := in["type"].(map[string]interface{}); ok && len(v) > 0 {
-		obj.Type = appsv1.DeploymentStrategyType(in["type"].(string))
+
+	if v, ok := in["type"]; ok {
```

Comment:
Probably we should check `in["type"]` if it is `string` and length of `v`
Like
```
if v, ok := in["type"].(string); ok && len(v) > 0 {
```

Reply:
This could be reduced to

```
if v, _ := in["type"].(string); v != "" {
```

Reply:
So one difference of checking `ok` instead of `v != ""` is in the scenario below:
```
strategy {
  type = "RollingUpdate"

  rolling_update {
    max_surge       = ""
    max_unavailable = ""
  }
}
```
You'll get this error message
```
kubernetes_deployment.scenario5: Failed to update deployment: Deployment.apps "scenario5" is invalid: [spec.strategy.rollingUpdate.maxUnavailable: Invalid value: intstr.IntOrString{Type:1, IntVal:0, StrVal:""}: a valid percent string must be a numeric string followed by an ending '%' (e.g. '1%',  or '93%', regex used for validation is '[0-9]+%'), spec.strategy.rollingUpdate.maxSurge: Invalid value: intstr.IntOrString{Type:1, IntVal:0, StrVal:""}: a valid percent string must be a numeric string followed by an ending '%' (e.g. '1%',  or '93%', regex used for validation is '[0-9]+%'), spec.strategy.rollingUpdate.maxUnavailable: Invalid value: intstr.IntOrString{Type:1, IntVal:0, StrVal:""}: may not be 0 when `maxSurge` is 0]
```

if we check `v != ""` then we'll silently fail to set the `max_surge` and `max_unavailable` which will cause subsequent applies to identify a change like below:
```
~ kubernetes_deployment.default
      spec.0.strategy.0.rolling_update.0.max_surge:       "25%" => ""
      spec.0.strategy.0.rolling_update.0.max_unavailable: "25%" => ""
```
I don't have a strong preference other than to be consistent with other resources.

What do you think?

Reply:
In this case `change` message can be more readable
```
~ kubernetes_deployment.default
      spec.0.strategy.0.rolling_update.0.max_surge:       "25%" => ""
      spec.0.strategy.0.rolling_update.0.max_unavailable: "25%" => ""
```
Will provide more info on why `change` occurs.